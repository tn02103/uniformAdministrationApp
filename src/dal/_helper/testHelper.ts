import dayjs from "@/lib/dayjs";


export const runServerActionTest = async <T>(functionPromise: Promise<T>): Promise<{ success: true, result: T } | { success: false, result: any }> => functionPromise
    .then((result) => {
        if ((result as any)?.error) {
            return {
                success: false,
                result: result
            }
        } else {
            return {
                success: true,
                result: result,
            }
        }
    }).catch((error) => ({
        success: false,
        result: error,
    }));

export type SAReturnType<T> = Promise<T | SAErrorType>

export type SAErrorType = {
    error: {
        message: string,
        formElement: string,
    }
}
export const compareDates = (stringDate: string, date: Date) => {
    return dayjs(stringDate).isSame(date, "day");
}

export const isToday = (date: string | Date) => {
    return dayjs.utc().isSame(date, "day");
}

export function cleanData(dataObject: any, attributesToRemove: string[]) {
    attributesToRemove.forEach(attribute => {
        const keys = attribute.split('.');
        removeAttribute(dataObject, keys);
    });
    return dataObject;
}

function removeAttribute(obj: any, keys: string[]) {
    if (keys.length === 0) return;
    if (obj === null || obj === undefined) return;

    const key = keys[0];
    if (keys.length === 1) {
        if (Array.isArray(obj)) {
            obj.forEach(item => delete item[key]);
        } else {
            delete obj[key];
        }
    } else {
        if (Array.isArray(obj)) {
            obj.forEach(item => removeAttribute(item[key], keys.slice(1)));
        } else if (obj[key]) {
            removeAttribute(obj[key], keys.slice(1));
        }
    }
}
