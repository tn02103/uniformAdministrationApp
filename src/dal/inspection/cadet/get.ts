import { genericSAValidator } from "@/actions/validations";
import { AuthRole } from "@/lib/AuthRoles";
import dayjs from "@/lib/dayjs";
import { prisma } from "@/lib/db";
import { CadetInspectionFormSchema } from "@/zod/deficiency";
import { z } from "zod";

/**
 * Fetches the inspection form data for a cadet.
 * 
 * @param props - The cadet ID.
 * @returns A promise that resolves to the inspection form data for the cadet.
 * @throws Will throw an error if no active inspection is found for today.
 */
export const getCadetInspectionFormData = async (props: string) => genericSAValidator(
    AuthRole.inspector,
    props,
    z.string().uuid(),
    { cadetId: props }
).then(async ([{ assosiation }, cadetId]) => {
    const activeInspection = await unsecuredGetActiveInspection(cadetId, assosiation);

    if (!activeInspection) {
        throw new Error("No active inspection found for today.");
    }

    const oldDeficiencies = await unsecuredGetPreviouslyUnresolvedDeficiencies(
        cadetId,
        activeInspection.fk_assosiation,
        activeInspection.id
    )

    const [issuedMaterials, issuedCounts, uniformTypes] = await prisma.$transaction([
        prisma.material.findMany({
            where: {
                issuedEntries: {
                    some: {
                        fk_cadet: cadetId,
                        dateReturned: null,
                    },
                },
            },
        }),
        prisma.uniform.groupBy({
            by: ['fk_uniformType'],
            where: {
                issuedEntries: {
                    some: {
                        fk_cadet: cadetId,
                        dateReturned: null,
                    },
                },
            },
            _count: true,
            orderBy: {
                fk_uniformType: 'asc',
            }
        }),
        prisma.uniformType.findMany({
            where: {
                fk_assosiation: assosiation,
                recdelete: null,
            }
        }),
    ]);
    const uniformComplete = uniformTypes.every(type => {
        const count = issuedCounts.find(count => count.fk_uniformType === type.id);
        if (!count || typeof count._count !== 'number')
             return false;

        return count._count >= type.issuedDefault;
    });


    return {
        cadetId,
        uniformComplete,
        oldDeficiencyList: oldDeficiencies.map(def => ({
            id: def.id,
            typeId: def.type.id,
            typeName: def.type.name,
            description: def.description,
            comment: def.comment,
            dateCreated: def.dateCreated,
            resolved: def.dateResolved !== null,
        })),
        newDeficiencyList: activeInspection.deficiencyCreated.map(def => {
            const isIssued = issuedMaterials.some(mat => mat.id === def.cadetDeficiency?.fk_material);

            return {
                id: def.id,
                typeId: def.type.id,
                description: def.description,
                comment: def.comment,
                uniformId: def.uniformDeficiency?.fk_uniform ?? def.cadetDeficiency?.fk_uniform ?? null,
                materialId: (isIssued ? def.cadetDeficiency?.fk_material : "other") ?? null,
                otherMaterialId: isIssued ? null : def.cadetDeficiency?.fk_material ?? null,
                otherMaterialGroupId: isIssued ? null : def.cadetDeficiency?.material?.fk_materialGroup ?? null,
                dateCreated: dayjs(def.dateCreated).format("YYYY-MM-DDTHH:mm:ss"),
            }
        }),
    } satisfies CadetInspectionFormSchema;
});

export const unsecuredGetPreviouslyUnresolvedDeficiencies = async (cadetId: string, assosiation: string, activeInspectionId: string) => {
    return prisma.deficiency.findMany({
        where: {
            type: { fk_assosiation: assosiation },
            AND: [
                {
                    OR: [
                        { cadetDeficiency: { fk_cadet: cadetId } },
                        {
                            uniformDeficiency: {
                                uniform: {
                                    issuedEntries: {
                                        some: {
                                            fk_cadet: cadetId,
                                            dateReturned: null,
                                        },
                                    },
                                },
                            },
                        },
                    ],
                },
                {
                    OR: [
                        { dateResolved: null },
                        { dateResolved: { not: null }, fk_inspection_resolved: activeInspectionId },
                    ],
                },
            ],
        },
        include: {
            type: true,
            cadetDeficiency: true,
            uniformDeficiency: true,
        },
        orderBy: [
            { dateCreated: 'asc'},
            { description: 'asc'}
        ]
    });
}
export const unsecuredGetActiveInspection = async (cadetId: string, assosiation: string) => {
    return await prisma.inspection.findFirst({
        where: {
            fk_assosiation: assosiation,
            date: dayjs().format("YYYY-MM-DD"),
            timeStart: { not: null },
            timeEnd: null,
        },
        include: {
            cadetInspection: {
                where: {
                    fk_cadet: cadetId,
                },
            },
            deficiencyCreated: {
                where: {
                    OR: [
                        { cadetDeficiency: { fk_cadet: cadetId } },
                        {
                            uniformDeficiency: {
                                uniform: {
                                    issuedEntries: {
                                        some: {
                                            fk_cadet: cadetId,
                                            dateReturned: null,
                                        },
                                    },
                                },
                            },
                        },
                    ],
                },
                include: {
                    type: true,
                    cadetDeficiency: {
                        include: {
                            material: true,
                        },
                    },
                    uniformDeficiency: true,
                },
            },
        },
    });
}
