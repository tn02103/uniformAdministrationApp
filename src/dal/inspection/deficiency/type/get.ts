import { genericSANoDataValidator } from "@/actions/validations";
import { AuthRole } from "@/lib/AuthRoles";
import { prisma } from "@/lib/db";
import { DeficiencyType, deficiencyTypeArgs } from "@/types/deficiencyTypes";

export const getUniformDefTypes = async (): Promise<DeficiencyType[]> => genericSANoDataValidator(
    AuthRole.inspector,
).then(([{ assosiation }]) => prisma.deficiencyType.findMany({
    where: {
        dependent: 'uniform',
        fk_assosiation: assosiation,
    },
    ...deficiencyTypeArgs
}));
