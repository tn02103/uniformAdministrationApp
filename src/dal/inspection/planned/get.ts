"use server";

import { genericSAValidatiorV2 } from "@/actions/validations";
import { AuthRole } from "@/lib/AuthRoles";
import { prisma } from "@/lib/db";
import { PlannedInspectionType } from "@/types/inspectionTypes";
import { PlannedInspectionDBQuery } from "./_dbQuerys";

const dbQuery = new PlannedInspectionDBQuery();
/**
 * Returns a list of all Inspections that are in the future, have not been active, or were not closed corectly.
 * @scope **
 * @returns list of Inspections
 */
export const getPlannedInspectionList = (): Promise<PlannedInspectionType[]> => genericSAValidatiorV2(
    AuthRole.materialManager, true, {}
).then(async ({ assosiation }) => dbQuery.plannedInspectionListQuery(assosiation, prisma));
