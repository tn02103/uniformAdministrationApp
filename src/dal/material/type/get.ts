import { genericSANoDataValidator } from "@/actions/validations";
import { AuthRole } from "@/lib/AuthRoles";
import { prisma } from "@/lib/db";
import { AdministrationMaterialGroup, materialGroupArgs } from "@/types/globalMaterialTypes";
import { Prisma } from "@prisma/client";

/**
 * Genneral configuration for Materials. Can be used in for all Authroles 
 * @returns Array of all MaterialGroups with Materials. MaterialGroups without any Materials are not included.
 */
export const getConfiguration = async () =>
    genericSANoDataValidator(AuthRole.user).then(async ([{ assosiation }]) =>
        prisma.materialGroup.findMany({
            ...materialGroupArgs,
            where: {
                fk_assosiation: assosiation,
                recdelete: null,
                typeList: {
                    some: {
                        recdelete: null,
                    }
                }
            },
            orderBy: { sortOrder: "asc" }
        })
    );


/**
 * Configuration of MaterialGroups & Types for Administration of theses. 
 * @returns list of MaterialGroups
 */
export const getAdministrationConfiguration = (): Promise<AdministrationMaterialGroup[]> =>
    genericSANoDataValidator(AuthRole.materialManager).then(async ([{ assosiation }]) => {
        const [groups, issuedQuantities] = await prisma.$transaction([
            getAdminList(assosiation),
            getMaterialIssueCountsByAssosiation(assosiation),
        ]);

        return groups.map(group => ({
            ...group,
            typeList: group.typeList.map(type => ({
                ...type,
                issuedQuantity: issuedQuantities.find(iq => iq.fk_material === type.id)?._sum.quantity ?? 0
            })),
        }));
    });

const getAdminList = (fk_assosiation: string, client?: Prisma.TransactionClient) =>
    (client ?? prisma).materialGroup.findMany({
        select: {
            id: true,
            description: true,
            sortOrder: true,
            issuedDefault: true,
            multitypeAllowed: true,
            typeList: {
                select: {
                    id: true,
                    typename: true,
                    actualQuantity: true,
                    targetQuantity: true,
                    sortOrder: true,
                },
                where: { recdelete: null },
                orderBy: { sortOrder: "asc" }
            },
        },
        where: { fk_assosiation, recdelete: null },
        orderBy: { sortOrder: "asc" }
    });
const getMaterialIssueCountsByAssosiation = (fk_assosiation: string) =>
    prisma.materialIssued.groupBy({
        by: ['fk_material'],
        _sum: { quantity: true },
        where: {
            material: {
                materialGroup: { fk_assosiation }
            },
            dateReturned: null,
        }
    });