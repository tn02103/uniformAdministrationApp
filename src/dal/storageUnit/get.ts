import { genericSANoDataValidator } from '@/actions/validations';
import { AuthRole } from '@/lib/AuthRoles';
import { prisma } from '@/lib/db';
import { uniformArgs } from '@/types/globalUniformTypes';
import { Prisma } from '@prisma/client';

const StorageUnitWithUniformItemsArgs = Prisma.validator<Prisma.StorageUnitFindManyArgs>()({
    include: {
        uniformList: {
            ...uniformArgs,
            orderBy: {
                number: 'asc',
            },
            where: {
                recdelete: null
            }
        },
    },
});

export type StorageUnitWithUniformItems = Prisma.StorageUnitGetPayload<typeof StorageUnitWithUniformItemsArgs>;

export const getUnitsWithUniformItems = (): Promise<StorageUnitWithUniformItems[]> => genericSANoDataValidator(AuthRole.user).then(
    ([{ assosiation }]) => __unsecuredGetUnitsWithUniformItems(assosiation),
);

export const __unsecuredGetUnitsWithUniformItems = (assosiation: string, client?: Prisma.TransactionClient): Promise<StorageUnitWithUniformItems[]> =>
    (client ?? prisma).storageUnit.findMany({
        where: { assosiationId: assosiation },
        ...StorageUnitWithUniformItemsArgs,
        orderBy: { name: 'asc' },
    });
