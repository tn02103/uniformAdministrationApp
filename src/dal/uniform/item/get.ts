import { genericSANoDataValidator, genericSAValidator } from "@/actions/validations";
import { AuthRole } from "@/lib/AuthRoles";
import { prisma } from "@/lib/db";
import { Deficiency } from "@/types/deficiencyTypes";
import { uniformHistoryArgs, UniformHistroyEntry, UniformWithOwner, uniformWithOwnerArgs } from "@/types/globalUniformTypes";
import { Prisma } from "@prisma/client";
import { z } from "zod";

export type ItemLabel = {
    id: string;
    label: string;
    active: boolean;
    typeId: string;
    number: number;
    owner: {
        id: string;
        firstname: string;
        lastname: string;
    } | null;
    storageUnit: {
        id: string;
        name: string;
    } | null;
}
/**
 * Get all uniform items for the assosiation
 * @requires AuthRole.user
 * @returns List of labels for uniform items
 */
export const getItemLabels = async (): Promise<ItemLabel[]> => genericSANoDataValidator(
    AuthRole.user,
).then(([{ assosiation }]) => prisma.uniform.findMany({
    where: {
        type: {
            fk_assosiation: assosiation
        },
        recdelete: null,
    },
    include: {
        type: true,
        storageUnit: true,
        issuedEntries: {
            where: {
                dateReturned: null,
            },
            include: {
                cadet: true,
            },
        },
    },
})).then(data => data.map((item): ItemLabel => ({
    id: item.id,
    number: item.number,
    label: `${item.type.name}-${item.number}`,
    active: item.active,
    typeId: item.type.id,
    owner: item.issuedEntries.length > 0 ? {
        id: item.issuedEntries[0].cadet.id,
        firstname: item.issuedEntries[0].cadet.firstname,
        lastname: item.issuedEntries[0].cadet.lastname,
    } : null,
    storageUnit: item.storageUnit ? {
        id: item.storageUnit.id,
        name: item.storageUnit.name,
    } : null,
})));


export const getHistory = async (props: string): Promise<UniformHistroyEntry[]> => genericSAValidator(
    AuthRole.inspector,
    props,
    z.string().uuid(),
    { uniformId: props }
).then(([, uniformId]) => prisma.uniformIssued.findMany({
    where: {
        fk_uniform: uniformId,
    },
    ...uniformHistoryArgs,
}));

const getDeficienciesPropSchema = z.object({
    uniformId: z.string().uuid(),
    includeResolved: z.boolean().optional(),
});
type GetDeficienciesProps = z.infer<typeof getDeficienciesPropSchema>;

export const getDeficiencies = async (props: GetDeficienciesProps): Promise<Deficiency[]> => genericSAValidator(
    AuthRole.inspector,
    props,
    getDeficienciesPropSchema,
    { uniformId: props.uniformId }
).then(([, { uniformId, includeResolved }]) => prisma.deficiency.findMany({
    where: {
        uniformDeficiency: {
            fk_uniform: uniformId,
        },
        dateResolved: includeResolved ? undefined : null,
    },
    include: {
        type: true,
        uniformDeficiency: true,
    },
    orderBy: [
        { dateCreated: 'asc' },  // Oldest to newest
    ],
})).then((deficiencies) =>
    deficiencies.map((deficiency): Deficiency => ({
        id: deficiency.id,
        typeId: deficiency.type.id,
        typeName: deficiency.type.name,
        description: deficiency.description,
        comment: deficiency.comment,
        dateCreated: deficiency.dateCreated,
        dateUpdated: deficiency.dateUpdated,
        dateResolved: deficiency.dateResolved,
        userCreated: deficiency.userCreated,
        userUpdated: deficiency.userUpdated,
        userResolved: deficiency.userResolved,
    }))
);

const getListWithOwnerPropSchema = z.object({
    uniformTypeId: z.string().uuid(),
    orderBy: z.enum(["number", "generation", "size", "comment", "owner"]),
    asc: z.boolean(),
    filter: z.object({
        active: z.boolean().optional(),
        isReserve: z.boolean().optional(),
        issued: z.boolean().optional(),
        notIssued: z.boolean().optional(),
        inStorageUnit: z.boolean().optional(),
        generations: z.record(z.boolean()).optional(),
        sizes: z.record(z.boolean()).optional(),
    }).optional().nullable(),
});
type getListWithOwnerProps = z.infer<typeof getListWithOwnerPropSchema>;

/**
 * used to get the data for the uniformList-Overview 
 * returns all data of unifrom, the owner is returned in the object of issuedEntries as a cadetDescription. 
 * @requires AuthRole.user
 * @param uniformTypeId 
 * @param orderBy number| generation | size | comment | owner
 * @param asc 
 * @param filter filterObject of uniformFilter Pannel
 * @returns UniformWithOwner[]
 */
export const getListWithOwner = async (props: getListWithOwnerProps): Promise<UniformWithOwner[]> => genericSAValidator(
    AuthRole.user,
    props,
    getListWithOwnerPropSchema,
    { uniformTypeId: props.uniformTypeId }
).then(async ([{ }, { uniformTypeId, orderBy, asc, filter }]) => {
    const sqlFilter: Prisma.UniformFindManyArgs["where"] = {};
    const generationFilter = filter?.generations ? Object.entries(filter.generations).filter(([, value]) => value).map(([key,]) => key) : [];
    const sizeFilter = filter?.sizes ? Object.entries(filter.sizes).filter(([, value]) => value).map(([key,]) => key) : [];
    const andConditions: Prisma.UniformWhereInput[] = [];
    if (!filter) {
        sqlFilter["active"] = true;
    } else {
        if (!filter.isReserve) {
            sqlFilter["active"] = true;
        } else if (!filter.active) {
            sqlFilter["active"] = false;
        }

        if (!filter.issued || !filter.notIssued || !filter.inStorageUnit) {
            const orConditions: Prisma.UniformWhereInput[] = [];
            if (filter.issued) {
                orConditions.push({
                    issuedEntries: {
                        some: {
                            dateReturned: null,
                        }
                    }
                });
            }
            if (filter.notIssued) {
                orConditions.push({
                    issuedEntries: {
                        none: {
                            dateReturned: null,
                        }
                    }
                });
            }
            if (filter.inStorageUnit) {
                orConditions.push({
                    storageUnit: {
                        isNot: null,
                    }
                });
            }
            andConditions.push({
                OR: orConditions,
            });
        }
    }
    if (generationFilter.length > 0) {
        andConditions.push({
            OR: [
                { generation: { id: { in: generationFilter } } },
                generationFilter.includes("null") ? { generation: null } : {}
            ]
        });
    }
    if (sizeFilter.length > 0) {
        andConditions.push({
            OR: [
                { size: { id: { in: sizeFilter } } },
                sizeFilter.includes("null") ? { size: null } : {}
            ]
        });
    }

    let sortOrder: Prisma.UniformOrderByWithRelationInput[];
    const ascString = asc ? "asc" : "desc";
    switch (orderBy) {
        case "generation":
            sortOrder = [
                { generation: { sortOrder: ascString } },
                { size: { sortOrder: ascString } },
                { number: ascString },
            ];
            break;
        case "size":
            sortOrder = [
                { size: { sortOrder: ascString } },
                { generation: { sortOrder: ascString } },
                { number: ascString },
            ];
            break;
        case "comment":
            sortOrder = [
                { comment: ascString },
                { number: ascString },
            ];
            break;
        case "owner":
            sortOrder = [
                { number: ascString }
            ];
            break;
        default:
            sortOrder = [{
                number: ascString,
            }];
            break;
    }

    const data = await prisma.uniform.findMany({
        ...uniformWithOwnerArgs,
        where: {
            ...sqlFilter,
            fk_uniformType: uniformTypeId,
            recdelete: null,
            AND: andConditions.length > 0 ? andConditions : undefined,
        },
        orderBy: sortOrder
    });

    if (orderBy !== "owner") {
        return data;
    }

    return data.sort((a, b) => {
        const returnValue = (value: number) => asc ? value : -value;

        // Sort by owner first
        if ((a.issuedEntries.length > 0) && (b.issuedEntries.length > 0)) {
            const nameA = a.issuedEntries[0].cadet.lastname + a.issuedEntries[0].cadet.firstname;
            const nameB = b.issuedEntries[0].cadet.lastname + b.issuedEntries[0].cadet.firstname;
            if (nameA === nameB) {
                return returnValue(a.number - b.number);
            } else {
                return returnValue(nameA.localeCompare(nameB));
            }
        }
        if (a.issuedEntries.length > 0) {
            return returnValue(-1);
        }
        if (b.issuedEntries.length > 0) {
            return returnValue(1);
        }
        // If no owner, sort by storage unit
        if (a.storageUnit && b.storageUnit) {
            if (a.storageUnit.name === b.storageUnit.name) {
                return returnValue(a.number - b.number);
            } else {
                return returnValue(a.storageUnit.name.localeCompare(b.storageUnit.name));
            }
        }
        if (a.storageUnit) {
            return returnValue(-1);
        }
        if (b.storageUnit) {
            return returnValue(1);
        }

        // If no owner and no storage unit, sort by number
        return returnValue(a.number - b.number);
    });
});


