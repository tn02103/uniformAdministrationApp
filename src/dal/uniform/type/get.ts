"use server"
import { genericSAValidator } from "@/actions/validations";
import { AuthRole } from "@/lib/AuthRoles";
import { prisma } from "@/lib/db";
import { uniformTypeArgs } from "@/types/globalUniformTypes";
import { z } from "zod";

export const getUniformType = (props: string) => genericSAValidator(
    AuthRole.user,
    props,
    z.string().uuid(),
).then(([typeId, { assosiation }]) => prisma.uniformType.findUnique({
    where: {
        id: typeId,
        fk_assosiation: assosiation,
    },
    ...uniformTypeArgs
}));


