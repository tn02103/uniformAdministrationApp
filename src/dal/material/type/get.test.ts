import { runServerActionTest } from "@/dal/_helper/testHelper";
import { prisma } from "@/lib/db";
import { AdministrationMaterialGroup, MaterialGroup } from "@/types/globalMaterialTypes";
import { StaticData } from "../../../../tests/_playwrightConfig/testData/staticDataLoader";
import { getMaterialAdministrationConfiguration, getMaterialConfiguration } from "./get";

const { ids, cleanup, data } = new StaticData(0);
describe('getMaterialConfiguration', () => {
    const getFunctionResult = () => runServerActionTest<MaterialGroup[]>(getMaterialConfiguration);
    it('should countain the right fields', async () => {
        const { success, result } = await getFunctionResult();
        expect(success).toBeTruthy();
        expect(result).toHaveLength(3);
        expect(Object.keys(result[0])).toStrictEqual([
            "id", "description", "issuedDefault", "multitypeAllowed", "sortOrder", "typeList"
        ]);
        expect(Object.keys(result[0].typeList[0])).toStrictEqual([
            "id", "typename", "sortOrder"
        ]);
        const groups = [data.materialGroups[0], data.materialGroups[1], data.materialGroups[2]];
        expect(result).toEqual(
            expect.arrayContaining(
                groups.map(
                    (group) => expect.objectContaining({
                        id: group.id,
                        issuedDefault: group.issuedDefault,
                        multitypeAllowed: group.multitypeAllowed,
                        description: group.description,
                        sortOrder: group.sortOrder,
                        typeList: expect.arrayContaining(
                            data.materialTypes
                                .filter(t => !t.recdelete && t.fk_materialGroup === group.id)
                                .map(
                                    type => expect.objectContaining({
                                        id: type.id,
                                        typename: type.typename,
                                        sortOrder: type.sortOrder,
                                    })
                                )
                        )
                    })
                )
            )
        );
    });
    it('should not contain deleted Groups', async () => getFunctionResult()
        .then(({ success, result }) => {
            expect(success).toBeTruthy();
            if (!success) return;
            expect(result).not.toEqual(
                expect.arrayContaining([
                    expect.objectContaining({
                        id: ids.materialGroupIds[3],
                    }),
                ]),
            );
        })
    );

    it('should not contain groups without Materials', async () => {
        try {
            await prisma.material.deleteMany({
                where: { fk_materialGroup: ids.materialGroupIds[4] }
            });
            const { success, result } = await getFunctionResult();
            expect(success).toBeTruthy();
            expect(result).not.toEqual(
                expect.arrayContaining([
                    expect.objectContaining({
                        id: ids.materialGroupIds[4],
                    }),
                ]),
            );
        } finally {
            await cleanup.materialConfig();
        }
    });
    it('should not contain groups with only deleted Materials', async () => getFunctionResult()
        .then(({ result, success }) => {
            expect(success).toBeTruthy();
            expect(result).not.toEqual(
                expect.arrayContaining([
                    expect.objectContaining({
                        id: ids.materialGroupIds[4],
                    }),
                ]),
            );
        })
    );
    it('both Material and Groups should be ordered by SortOrder', async () => {
        const { success, result } = await getFunctionResult();
        expect(success).toBeTruthy();
        if (!success) return;
        expect(result.map(g => g.id)).toEqual(
            [ids.materialGroupIds[0], ids.materialGroupIds[1], ids.materialGroupIds[2]]
        );
        expect(result[0].typeList.map(t => t.id)).toEqual(
            [ids.materialIds[0], ids.materialIds[1], ids.materialIds[2], ids.materialIds[3]]
        );
    })
    it('should not contain deleted Materials', async () => {
        const { result, success } = await getFunctionResult();
        expect(success).toBeTruthy();
        expect(result[1].id).toBe(ids.materialGroupIds[1]);
        expect(result[1].typeList).not.toBe(
            expect.arrayContaining([
                expect.objectContaining({
                    id: ids.materialIds[10]
                }),
            ]),
        );
    });
});

describe('getMaterialAdministrationConfiguration', () => {
    const getFunctionResult = () => runServerActionTest<AdministrationMaterialGroup[]>(getMaterialAdministrationConfiguration);
    it('should countain the right fields', async () => {
        const { success, result } = await getFunctionResult();
        expect(success).toBeTruthy();
        expect(result).toHaveLength(4);
        expect(Object.keys(result[0])).toStrictEqual(
            expect.arrayContaining([
                "id", "description", "issuedDefault", "multitypeAllowed", "sortOrder", "typeList"
            ])
        );
        expect(Object.keys(result[0].typeList[0])).toStrictEqual(
            expect.arrayContaining([
                "id", "typename", "sortOrder", "actualQuantity", "targetQuantity", "issuedQuantity"
            ])
        );
        const groups = [data.materialGroups[0], data.materialGroups[1], data.materialGroups[2], data.materialGroups[4]];
        expect(result).toEqual(
            expect.arrayContaining(
                groups.map(
                    (group) => expect.objectContaining({
                        id: group.id,
                        issuedDefault: group.issuedDefault,
                        multitypeAllowed: group.multitypeAllowed,
                        description: group.description,
                        sortOrder: group.sortOrder,
                        typeList: expect.arrayContaining(
                            data.materialTypes
                                .filter(t => !t.recdelete && t.fk_materialGroup === group.id)
                                .map(
                                    type => expect.objectContaining({
                                        id: type.id,
                                        typename: type.typename,
                                        sortOrder: type.sortOrder,
                                        actualQuantity: type.actualQuantity,
                                        targetQuantity: type.targetQuantity,
                                    })
                                )
                        )
                    })
                )
            )
        );
    });
    it('should not contain deleted Groups', async () => getFunctionResult()
        .then(({ success, result }) => {
            expect(success).toBeTruthy();
            if (!success) return;
            expect(result).not.toEqual(
                expect.arrayContaining([
                    expect.objectContaining({
                        id: ids.materialGroupIds[3],
                    }),
                ]),
            );
        })
    );

    it('should contain groups without Materials', async () => {
        try {
            await prisma.material.deleteMany({
                where: { fk_materialGroup: ids.materialGroupIds[4] }
            });
            const { success, result } = await getFunctionResult();
            expect(success).toBeTruthy();
            expect(result).toEqual(
                expect.arrayContaining([
                    expect.objectContaining({
                        id: ids.materialGroupIds[4],
                    }),
                ]),
            );
        } finally {
           await cleanup.materialConfig();
        }
    });
    it('should contain groups with only deleted Materials', async () => getFunctionResult()
        .then(({ result, success }) => {
            expect(success).toBeTruthy();
            expect(result).toEqual(
                expect.arrayContaining([
                    expect.objectContaining({
                        id: ids.materialGroupIds[4],
                    }),
                ]),
            );
        })
    );
    it('both Material and Groups should be ordered by SortOrder', async () => {
        const { success, result } = await getFunctionResult();
        expect(success).toBeTruthy();
        if (!success) return;
        expect(result.map(g => g.id)).toEqual(
            [ids.materialGroupIds[0], ids.materialGroupIds[1], ids.materialGroupIds[2], ids.materialGroupIds[4]]
        );
        expect(result[0].typeList.map(t => t.id)).toEqual(
            [ids.materialIds[0], ids.materialIds[1], ids.materialIds[2], ids.materialIds[3]]
        );
    })
    it('should not contain deleted Materials', async () => {
        const { result, success } = await getFunctionResult();
        expect(success).toBeTruthy();
        expect(result[1].id).toBe(ids.materialGroupIds[1]);
        expect(result[1].typeList).not.toBe(
            expect.arrayContaining([
                expect.objectContaining({
                    id: ids.materialIds[10]
                }),
            ]),
        );
    });
});
