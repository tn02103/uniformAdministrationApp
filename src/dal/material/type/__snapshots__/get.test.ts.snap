// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getMaterialAdministrationConfiguration, validate snapshot 1`] = `
[
  {
    "description": "Gruppe1",
    "issuedDefault": null,
    "multitypeAllowed": false,
    "sortOrder": 0,
    "typeList": [
      {
        "actualQuantity": 200,
        "issuedQuantity": 13,
        "sortOrder": 0,
        "targetQuantity": 150,
        "typename": "Typ1-1",
      },
      {
        "actualQuantity": 300,
        "issuedQuantity": 0,
        "sortOrder": 1,
        "targetQuantity": 0,
        "typename": "Typ1-2",
      },
      {
        "actualQuantity": 100,
        "issuedQuantity": 1,
        "sortOrder": 2,
        "targetQuantity": 200,
        "typename": "Typ1-3",
      },
      {
        "actualQuantity": 1,
        "issuedQuantity": 3,
        "sortOrder": 3,
        "targetQuantity": 20,
        "typename": "Typ1-4",
      },
    ],
  },
  {
    "description": "Gruppe2",
    "issuedDefault": 4,
    "multitypeAllowed": false,
    "sortOrder": 1,
    "typeList": [
      {
        "actualQuantity": 200,
        "issuedQuantity": 8,
        "sortOrder": 0,
        "targetQuantity": 200,
        "typename": "Typ2-1",
      },
      {
        "actualQuantity": 200,
        "issuedQuantity": 24,
        "sortOrder": 1,
        "targetQuantity": 200,
        "typename": "Typ2-2",
      },
      {
        "actualQuantity": 200,
        "issuedQuantity": 11,
        "sortOrder": 2,
        "targetQuantity": 200,
        "typename": "Typ2-3",
      },
    ],
  },
  {
    "description": "Gruppe3",
    "issuedDefault": null,
    "multitypeAllowed": true,
    "sortOrder": 2,
    "typeList": [
      {
        "actualQuantity": 0,
        "issuedQuantity": 8,
        "sortOrder": 0,
        "targetQuantity": 0,
        "typename": "Typ3-1",
      },
      {
        "actualQuantity": 0,
        "issuedQuantity": 9,
        "sortOrder": 1,
        "targetQuantity": 0,
        "typename": "Typ3-2",
      },
      {
        "actualQuantity": 0,
        "issuedQuantity": 12,
        "sortOrder": 2,
        "targetQuantity": 0,
        "typename": "Typ3-3",
      },
    ],
  },
  {
    "description": "Gruppe5",
    "issuedDefault": null,
    "multitypeAllowed": false,
    "sortOrder": 3,
    "typeList": [],
  },
]
`;

exports[`getMaterialConfiguration, validate snapshot 1`] = `
[
  {
    "description": "Gruppe1",
    "issuedDefault": null,
    "multitypeAllowed": false,
    "sortOrder": 0,
    "typeList": [
      {
        "sortOrder": 0,
        "typename": "Typ1-1",
      },
      {
        "sortOrder": 1,
        "typename": "Typ1-2",
      },
      {
        "sortOrder": 2,
        "typename": "Typ1-3",
      },
      {
        "sortOrder": 3,
        "typename": "Typ1-4",
      },
    ],
  },
  {
    "description": "Gruppe2",
    "issuedDefault": 4,
    "multitypeAllowed": false,
    "sortOrder": 1,
    "typeList": [
      {
        "sortOrder": 0,
        "typename": "Typ2-1",
      },
      {
        "sortOrder": 1,
        "typename": "Typ2-2",
      },
      {
        "sortOrder": 2,
        "typename": "Typ2-3",
      },
    ],
  },
  {
    "description": "Gruppe3",
    "issuedDefault": null,
    "multitypeAllowed": true,
    "sortOrder": 2,
    "typeList": [
      {
        "sortOrder": 0,
        "typename": "Typ3-1",
      },
      {
        "sortOrder": 1,
        "typename": "Typ3-2",
      },
      {
        "sortOrder": 2,
        "typename": "Typ3-3",
      },
    ],
  },
]
`;
